### ISSUE TITLE
3D plot surfaces of calibration error vs initial estimate don't make sense

### ISSUE THREAD

**Issue by danifpdra** [lardemua/atom#281]
Hi @miguelriemoliveira,

I used batch execution to calibrate the mmtbot setup with initial estimates from 0-2m in translation and 0-40 degrees in rotation. I don't think these results make much sense. Can you tell me your opinion?

![image](https://user-images.githubusercontent.com/47574905/114579096-70a68500-9c75-11eb-92c0-72178a906d27.png)
![image](https://user-images.githubusercontent.com/47574905/114579109-73a17580-9c75-11eb-961a-67897e05fe87.png)
![image](https://user-images.githubusercontent.com/47574905/114579122-769c6600-9c75-11eb-86a0-b8abccb3ff21.png)
![image](https://user-images.githubusercontent.com/47574905/114579140-79975680-9c75-11eb-8a18-c1d11a4f0f6e.png)

**Comment by danifpdra** [comment]
Hi @miguelriemoliveira,

The batch execution got stuck on 

`rosrun atom_calibration calibrate -json /home/daniela/datasets/train_dataset_1/data_collected.json  -csf 'lambda name: int(name)<1' -nig 0.10 0.30 -ss 1`

for like 5 hours. I don't think this is very normal. I've tried to run the same command separately and it finished really quickly.

**Comment by danifpdra** [comment]
New data: 

![image](https://user-images.githubusercontent.com/47574905/114711692-2bdb2680-9d27-11eb-83a8-ad3e0f512275.png)
![image](https://user-images.githubusercontent.com/47574905/114711708-309fda80-9d27-11eb-9fb5-e4a76558114e.png)
![image](https://user-images.githubusercontent.com/47574905/114711723-34336180-9d27-11eb-95c8-d3d64de8494c.png)
![image](https://user-images.githubusercontent.com/47574905/114711733-372e5200-9d27-11eb-875d-152ccb5d3c68.png)

**Comment by danifpdra** [comment]
More resultion: distance and angle 0.5 to 0.5, 21 points each.

I can only run the batch execution with the calibrate file options like this:

`options = {'ftol': 1e-4, 'xtol': 1e-4, 'gtol': 1e-4, 'diff_step': None, 'x_scale': 'jac'}`

@miguelriemoliveira , do you have any idea why?

![image](https://user-images.githubusercontent.com/47574905/114716314-d9e8cf80-9d2b-11eb-96ab-f7fa474a3630.png)
![image](https://user-images.githubusercontent.com/47574905/114716326-dc4b2980-9d2b-11eb-92f0-f9a69e717410.png)
![image](https://user-images.githubusercontent.com/47574905/114716344-e0774700-9d2b-11eb-84b2-06f9d8e54dec.png)
![image](https://user-images.githubusercontent.com/47574905/114716357-e4a36480-9d2b-11eb-9c20-ab83572fe494.png)

**Comment by danifpdra** [comment]
Even more resolution. From 0 to 0.75, 50 points each 

![image](https://user-images.githubusercontent.com/47574905/114738133-9ac47980-9d3f-11eb-9dc9-d518ce8d2461.png)
![image](https://user-images.githubusercontent.com/47574905/114738147-9e580080-9d3f-11eb-933b-4ae1d8de8894.png)
![image](https://user-images.githubusercontent.com/47574905/114738162-a31cb480-9d3f-11eb-8f86-e65a2156a211.png)
![image](https://user-images.githubusercontent.com/47574905/114738175-a6b03b80-9d3f-11eb-9ba8-20acb9cfa497.png)

**Comment by danifpdra** [comment]
Hi @miguelriemoliveira ,

The file to generate the plots is here: https://github.com/miguelriemoliveira/mmtbot/blob/c8f9ba72c1b2e50a737eea8e8d39cae80ee515a9/mmtbot_calibration/scripts/mmtbot%20analysis.ipynb

**Comment by danifpdra** [comment]
This file also generates the commands to a txt file that you can copy to the batch file. You just need to choose the initial and final value and the number of points for each vector.

**Comment by miguelriemoliveira** [comment]
> Hi @miguelriemoliveira,
> 
> The batch execution got stuck on
> 
> `rosrun atom_calibration calibrate -json /home/daniela/datasets/train_dataset_1/data_collected.json -csf 'lambda name: int(name)<1' -nig 0.10 0.30 -ss 1`
> 
> for like 5 hours. I don't think this is very normal. I've tried to run the same command separately and it finished really quickly.

Hi @danifpdra ,

I am trying to debug this,  but its very hard because your **batch_mmtbot,yml** is very large, its running now for 10 minutes and still going...

By the way, you did not say but I assume the command is:

`./batch_execution -v -f batch_mmtbot.yml`

My question is: do you have a smaller example where it crashes? I mean, a smaller batch config file? Otherwise its nearly impossible to debug if every new try takes 10 or 15 minutes...

**Comment by miguelriemoliveira** [comment]
I think I finally got stuck here:

```
Batch38, executing command:
rosrun atom_calibration calibrate -json /home/mike/datasets/mmtbot/train_dataset/data_collected.json  -csf 'lambda name: int(name)<1' -nig 0.00 0.58 -ss 1
```

The parameters are not the ones you have, so I suspect it has nothing to do with the command parameters, rather with something in the batch...

and the processors fall to 0% (before they were at 100%), as if the code just freezes or halts ... is this similar to what occurs with you?

**Comment by danifpdra** [comment]
Yes, this is what happens to me. But I have run a batch with 2500 calibration and it worked well as you have seen so I don't know what is happening.

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

I think I may have solved it. Can you try it please?

Also, I could try to make a batch execution which would run commands in parallel, but I am not really sure if this is very high priority. What do you think?

**Comment by danifpdra** [comment]
I will try it, yes.

I don't think it is important. It is working the way it is, just might take longer but it's no big deal.

**Comment by miguelriemoliveira** [comment]
OK, lets forget about it for now. About the fix, I am now running trial 69 of a 10-8 setting and going fine ...

**Comment by danifpdra** [comment]
I will try to run the big calibration with the 10e-8 and see what happens. That way we can even see the difference in the surfaces of that.

**Comment by miguelriemoliveira** [comment]
Right now I am running trial 453 and still going. I think that problem is solved.

**Comment by danifpdra** [comment]
Hi, plots for the same distances and angles with 10e-8:

![image](https://user-images.githubusercontent.com/47574905/114948095-47891e80-9e46-11eb-9680-04f62b3eb2e7.png)
![image](https://user-images.githubusercontent.com/47574905/114948106-4b1ca580-9e46-11eb-8ad2-d4a86ea5df07.png)
![image](https://user-images.githubusercontent.com/47574905/114948109-4e179600-9e46-11eb-8a04-04a7da2de615.png)
![image](https://user-images.githubusercontent.com/47574905/114948119-51128680-9e46-11eb-8734-10f3b0f5f499.png)

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

tell me your opinion about these new results.

Mine is that they look good. I would say no need to get more results of this kind...

I will try to beautify the graphs a bit more and then ask for your opinion.

**Comment by danifpdra** [comment]
Ok. I've updated the code for plotting these surfaces: https://github.com/miguelriemoliveira/mmtbot/blob/c8f9ba72c1b2e50a737eea8e8d39cae80ee515a9/mmtbot_calibration/scripts/mmtbot%20analysis.ipynb.

**Comment by miguelriemoliveira** [comment]
Ok, thanks, but the new version generates the images above?

**Comment by danifpdra** [comment]
Yes, it does.

**Comment by eupedrosa** [comment]
Wouldn't it be better to use something like a heat map? Or contour map ?

Although these graphics are interesting,  and somewhat expeted. Their non-smothness is not very appealling.

**Comment by miguelriemoliveira** [comment]
Yes, also something else: we are using two degrees of freedom (height and color) to display a single variable.

@danifpdra , can you try something like the matriz de risco that we see for the pandemic situation? Basically a top down view.

Also, the colors should perhaps be more intuitive: a green (good) to red(bad) colormap perhaps is more suited?

**Comment by danifpdra** [comment]
Is this better?

![image](https://user-images.githubusercontent.com/47574905/115221502-a75f1e00-a101-11eb-9d70-88751aace367.png)

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

much better! Nice work.

For a paper in a top journal we still need to improve the following:

What do the axes represent (including units)? 

I like the one on the left, but the xtick tics should perhaps be similar. You have 5 values (0 0.2 0.4 0.6 and 0.8 ) in x and many more on y. Can you fix it?

The figure on the right requires a legend that explains what is green and what is red.

Can you improve with these tips? Think about a top publication and how every little detail should be looked after.
When you have new graphs then we can discuss and select one of them to insert in the paper.

**Comment by danifpdra** [comment]
Is this better? Do you prefer the matlab (previous plot) or python (this one) plots ?

![image](https://user-images.githubusercontent.com/47574905/115262098-21a49800-a12c-11eb-8921-738531da14e1.png)

**Comment by danifpdra** [comment]
This is the python version

![image](https://user-images.githubusercontent.com/47574905/115262575-95df3b80-a12c-11eb-98b1-9245998cbc28.png)
![image](https://user-images.githubusercontent.com/47574905/115262598-9a0b5900-a12c-11eb-9f32-b46dece46624.png)
![image](https://user-images.githubusercontent.com/47574905/115262629-9e377680-a12c-11eb-9055-b6a27873fd25.png)
![image](https://user-images.githubusercontent.com/47574905/115262651-a2fc2a80-a12c-11eb-9012-ddfbacdf8184.png)

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

this is getting much better. I would say python, since its completely integrated into atom.

Some more comments:

In the paper the title of the graphs should be removed.
Initial estimate distance -> Linear distance from initial estimate to ground truth (meters)
Initial estimate angle -> Angular distance from initial estimate to ground truth (degrees)
Time -> Appears on top of the zticks
Time -> Time (secs)
World camera error -> Error (pixels) -> the legend will refer which camera the graph is for
Hand camera error -> Error (pixels) -> the legend will refer which camera the graph is for
Lidar camera error -> Error (meters) -> the legend will refer which sensor the graph is for
grey color behind the plot, the background, can it be white?

I like the graphs on the left more, and you?

**Comment by danifpdra** [comment]
Is this what you meant?

Yes, I agree.

![image](https://user-images.githubusercontent.com/47574905/115274298-69312100-a138-11eb-931a-6dc471e72a4a.png)

**Comment by miguelriemoliveira** [comment]
Yes, looks much better than at the start of the issue. Congratullations.

I think the axis labels have the correct text now but its too large, so its better to add symbols and then refer to them in the figure caption. 

So for example you can use

l (meters) and then in the caption explain what l is
a (degrees) and then in the caption explain what a is

But it looks great. I would move it into the paper.

**Comment by danifpdra** [comment]
I inserted the plots in the paper. Can you check if it's alright?

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

hum, not just yet, because the vertical axis legend is colliding with the ztick values.

Also, for reference, never have in a graph text which is smaller that the smallest text in the body of the document, which is normally the figure captions or the footnotes. That means you will have to increase the size of the font of the ticks and the legend a bit.

The colorbar with the legend is not needed because its redundant with the information in the z axis ticks.

![image](https://user-images.githubusercontent.com/11350998/115376499-a6dc8b00-a1c6-11eb-804c-ae860a763fbf.png)

You are going to kill me, but remember we are aiming at the best of the best journals.

**Comment by danifpdra** [comment]
Is this ok?


![image](https://user-images.githubusercontent.com/47574905/115382691-de4e3600-a1cc-11eb-9d2a-10705da004ab.png)

**Comment by danifpdra** [comment]
Is this better?

![image](https://user-images.githubusercontent.com/47574905/115383275-87952c00-a1cd-11eb-9333-f7ad21105541.png)

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra ,

yes, definitely. I don't think the colorbar is needed. I would remove.
The values in Angular distance are strange .. they are not in degrees, are they?

Also, the values of the error are strange. Are you sure those are pixels? Are they not meters? If so perhaps consider using millimeters in the graph for a better visualization.

**Comment by danifpdra** [comment]
You are right,  the angular distance is in rad, not degrees. Do you want me to convert to degrees or should I just change the labels?

Yes, that particular plot is from LiDAR, the errors are in mm. Sorry, I copy-pasted and didn't notice it. I have already changed it.

**Comment by miguelriemoliveira** [comment]
Ahah ... everytime I catch one more typo in the graph you owe me a coffee :).

I would prefer degrees, its more intuitive

**Comment by danifpdra** [comment]
Did I get it right this time?

![image](https://user-images.githubusercontent.com/47574905/115406547-31cd7d80-a1e7-11eb-9b11-038003521abe.png)

**Comment by miguelriemoliveira** [comment]
No : - ), the z axis values are in meters and you are saying in the legend that they are millimeters. 

Can you multiply those values by 1000?

Do I get my coffee : - )?

**Comment by danifpdra** [comment]
I'm confident that this is still not correct :-)

![image](https://user-images.githubusercontent.com/47574905/115472963-bdb8c700-a232-11eb-91d9-b1e34f7704a0.png)

**Comment by miguelriemoliveira** [comment]
Its great. Thanks for the patience : - ). I hope you agree that now its much better than at the beginning ...

And don't forget my coffee!

**Comment by danifpdra** [comment]
Yes, I agree. Thanks!

It's already on its way, I sent it on high priority mail :-)

**Comment by miguelriemoliveira** [comment]
Hi @danifpdra . Can we close this issue? Its done, right?
